File: PythonLeetcode/Leetcode/730_CountDifferentPalindromicSubsequences.py
"""
730. Count Different Palindromic Subsequences

# Given a string S, find the number of different non-empty palindromic subsequences in S, and return that number modulo 10^9 + 7.

# A subsequence of a string S is obtained by deleting 0 or more characters from S.

# A sequence is palindromic if it is equal to the sequence reversed.

# Two sequences A_1, A_2, ... and B_1, B_2, ... are different if there is some i for which A_i != B_i.

# Example 1:
# Input:
# S = 'bccb'
# Output: 6
# Explanation:
# The 6 different non-empty palindromic subsequences are 'b', 'c', 'bb', 'cc', 'bcb', 'bccb'.
# Note that 'bcb' is counted only once, even though it occurs twice.

# Example 2:
# Input:
# S = 'abcdabcdabcdabcdabcdabcdabcdabcddcbadcbadcbadcbadcbadcbadcbadcba'
# Output: 104860361
# Explanation:
# There are 3104860382 different non-empty palindromic subsequences, which is 104860361 modulo 10^9 + 7.

# Note:

# The length of S will be in the range [1, 1000].
# Each character S[i] will be in the set {'a', 'b', 'c', 'd'}.
"""


class CountPalindromicSubsequences:

    """
    # I started out with only the DFS part, got TLE, then designed the DP part cache and passed.

    # The computations are done in DFS(start, end) in which I computed the answer for the string S[start, end].
    # cache is only to save the results of the computations in order to avoid repetitions.

    # In DFS(start, end), for instance, for the letter 'a',
    # I compute the number of palindromes that start and end with 'a' in the following way:

    # First of all, I compute when 'a' appears first (index i) and last (index j) in the segment I am considering.
    # Then it breaks down into two cases:

    # If i == j. There is only one 'a' in the segment. So the answer is 1.

    # If i != j. The possible palindromes are 'a', 'aa', and 'a*a' where '*' stands for any palindromes contained in S[i+1:j].
    # The answer would be DFS(i+1,j) + 2. Since I want to avoid repetitive computation, I write cache(i+1,j) + 2 instead.

    # The worst case time complexity is O(n^2). The best case time complexity is O(n).

    # Btw, to make this algorithm even faster, one could set check to be a 2D list instead of a dictionary,
    # but that would occupy more space.
    """
    def doit_dp_dfs(self, S):
        """
        :type S: str
        :rtype: int
        """
        dp = {}

        # return the amount of distinct palindromes of S[start:end]
        def search(start, end):

            if (start, end) in dp:
                return dp[(start, end)]

            # if less then two, 'bb', 'bc' are two elements
            if end <= start + 2:
                return end - start

            count = 0
            for c in ['a', 'b', 'c', 'd']:

                # s, e = S.index(c, start, end), S.rindex(c, start, end)
                s, e = S.find(c, start, end), S.rfind(c, start, end)

                if s == -1 or e == -1:
                    continue

                # could be 'a' 'aa' and searched 'a*a'
                count += search(s+1, e) + 2 if s != e else 1

            dp[(start, end)] = count % 1000000007

            return dp[(start, end)]
            
        return search(0, len(S))

    def doit_dfs_dp(self, S: str) -> int:
        from functools import lru_cache
        mod = 10 ** 9 + 7
        chars = set(S)
        next = {key: [len(S)] * len(S) for key in chars}
        prev = {key: [-1] * len(S) for key in chars}

        for key in chars:
            prev_position = -1
            for i in range(len(S)):
                if S[i] == key:
                    prev_position = i
                prev[key][i] = prev_position

            next_position = len(S)
            for i in range(len(S) - 1, -1, -1):
                if S[i] == key:
                    next_position = i
                next[key][i] = next_position

        @lru_cache(None)
        def search(i, j):
            if j < i:
                return 0
            # count number of palindromic sequence in S[i: j + 1]
            res = 0
            for key in chars:
                if j < next[key][i]:
                    continue

                if next[key][i] == prev[key][j]:
                    res += 1
                    continue

                # next[key][i] < prev[key][j]
                res += 2 + search(next[key][i] + 1, prev[key][j] - 1)
            return res % mod

        return search(0, len(S) - 1) % mod

    """
    Approach #1 Dynamic Programming (using 3D array) [Accepted]
    Intuition and Algorithm
    
    Let dp[x][i][j] be the answer for the substring S[i...j] where S[i] == S[j] == 'a'+x. Note that since we only have 4 characters a, b, c, d, thus 0 <= x < 4. The DP formula goes as follows:
    
    If S[i] != 'a'+x, then dp[x][i][j] = dp[x][i+1][j], note that here we leave the first character S[i] in the window out due to our definition of dp[x][i][j].
    
    If S[j] != 'a'+x, then dp[x][i][j] = dp[x][i][j-1], leaving the last character S[j] out.
    
    If S[i] == S[j] == 'a'+x, then dp[x][i][j] = 2 + dp[0][i+1][j-1] + dp[1][i+1][j-1] + dp[2][i+1][j-1] + dp[3][i+1][j-1]. 
    When the first and last characters are the same, we need to count all the distinct palindromes (for each of a,b,c,d) within the sub-window S[i+1][j-1] plus the 2 palindromes contributed by the first and last characters.
    
    Let n be the length of the input string S, The final answer would be dp[0][0][n-1] + dp[1][0][n-1] + dp[2][0][n-1] + dp[3][0][n-1] mod 1000000007.


    """
    def doit_dp_2(self, S):
        n = len(S)
        mod = 1000000007
        dp = [[[0 for _ in range(n)] for _ in range(n)] for _ in range(4)]

        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                for k in range(4):
                    c = chr(ord('a') + k)
                    if j == i:
                        if S[i] == c:
                            dp[k][i][j] = 1
                        else:
                            dp[k][i][j] = 0
                    else:  # j > i
                        if S[i] != c:
                            dp[k][i][j] = dp[k][i + 1][j]
                        elif S[j] != c:
                            dp[k][i][j] = dp[k][i][j - 1]
                        else:  # S[i] == S[j] == c
                            if j == i + 1:
                                dp[k][i][j] = 2  # "aa" : {"a", "aa"}
                            else:  # length is > 2
                                dp[k][i][j] = 2
                                for m in range(4):  # count each one within subwindows [i+1][j-1]
                                    dp[k][i][j] += dp[m][i + 1][j - 1]
                                    dp[k][i][j] %= mod

        ans = 0
        for k in range(4):
            ans += dp[k][0][n - 1]
            ans %= mod

        return ans


    """
    Approach #2: Dynamic Programming (using 2D array) [Accepted]
    Intuition
    
    Almost every palindrome is going to take one of these four forms: a_a, b_b, c_c, or d_d, where _ represents a palindrome of zero or more characters. (The only other palindromes are a, b, c, d, and the empty string.)
    
    Let's try to count palindromes of the form a_a - the other types are similar. Evidently, we should take the first and last a, 
    then count all the palindromes that can be formed in between, as this provides us strictly more possibilities for _ to be a palindrome. This reveals an optimal substructure that is ideal for dynamic programming.
    
    Algorithm
    
    Let dp(i, j) be the number of palindromes (including the palindrome '') in the string T = S[i], S[i+1], ..., S[j]. To count palindromes in T of the form a_a, 
    we will need to know the first and last occurrence of 'a' in this string. This can be done by a precomputed dp: next[i][0] will be the next occurrence of 'a' in S[i:], next[i][1] will be the next occurrence of 'b' in S[i:], and so on.
    
    Also, we will need to know the number of unique letters in T to count the single letter palindromes. We can use the information from next to deduce it: if next[i][0] is in the interval [i, j], then 'a' occurs in T, and so on.
    
    As many states dp(i, j) do not need to be computed, the most natural approach is a top-down variation of dynamic programming.
    """
    def doit_dp_1(self, S):
        N = len(S)
        A = [ord(c) - ord('a') for c in S]
        prv = [None] * N
        nxt = [None] * N

        last = [None] * 4
        for i in xrange(N):
            last[A[i]] = i
            prv[i] = tuple(last)

        last = [None] * 4
        for i in range(N-1, -1, -1):
            last[A[i]] = i
            nxt[i] = tuple(last)

        MOD = 10**9 + 7
        memo = [[None] * N for _ in range(N)]
        def dp(i, j):
            if memo[i][j] is not None:
                return memo[i][j]
            ans = 1
            if i <= j:
                for x in range(4):
                    i0 = nxt[i][x]
                    j0 = prv[j][x]
                    if i <= i0 <= j:
                        ans += 1
                    if None < i0 < j0:
                        ans += dp(i0+1, j0-1)
            ans %= MOD
            memo[i][j] = ans
            return ans

        return dp(0, N-1) - 1


    def doit1(self, S):
        """
        :type S: str
        :rtype: int
        """
        result = set()
        d = set()
        
        def search(content, prefix, suffix,  result):

            if content in d:
                return 
            if not content:
                return 

            for i in set(content):
                s, e = content.find(i), content.rfind(i)
                result.add( prefix + i + suffix)
                if s != e:
                    result.add( prefix + i + i + suffix)
                    search(content[s+1: e], prefix+i, i+suffix, result)
            d.add(content)
            return 
                    
        search(S, '', '', result)
        return len(result)
        

if __name__ == "__main__":

    res = CountPalindromicSubsequences().doit_dp_dfs('bccb')

    #res = CountPalindromicSubsequences().doit_dp_dfs('baccdb')

    #res = CountPalindromicSubsequences().doit_dp_dfs('abcdabcdabcdabcdabcdabcdabcdabcddcbadcbadcbadcbadcbadcbadcbadcba')

File: cppLeetcode/leetcode/730_CountDifferentPalindromicSubsequences.cpp
/*
 
 730. Count Different Palindromic Subsequences
 
 Given a string S, find the number of different non-empty palindromic subsequences in S, and return that number modulo 10^9 + 7.

 A subsequence of a string S is obtained by deleting 0 or more characters from S.

 A sequence is palindromic if it is equal to the sequence reversed.

 Two sequences A_1, A_2, ... and B_1, B_2, ... are different if there is some i for which A_i != B_i.

 Example 1:
 Input:
 S = 'bccb'
 Output: 6
 Explanation:
 The 6 different non-empty palindromic subsequences are 'b', 'c', 'bb', 'cc', 'bcb', 'bccb'.
 Note that 'bcb' is counted only once, even though it occurs twice.
 Example 2:
 Input:
 S = 'abcdabcdabcdabcdabcdabcdabcdabcddcbadcbadcbadcbadcbadcbadcbadcba'
 Output: 104860361
 Explanation:
 There are 3104860382 different non-empty palindromic subsequences, which is 104860361 modulo 10^9 + 7.
 Note:

 The length of S will be in the range [1, 1000].
 Each character S[i] will be in the set {'a', 'b', 'c', 'd'}.

 */
#include <string>
#include <vector>

using std::vector;
using std::string;


class CountPalindromicSubsequences {

    /*
        730.Count-Different-Palindromic-Subsequences
        我们定义dp[i][j]为区间[i,j]内的不同的回文序列的个数。那么怎么区分不同的回文序列呢？可以采用层层剥离的思想，也就是分别考虑回文序列的最外层配对是'a','b','c','d'的四种情况.按照这种思路去分类的话，不同类之间的回文序列肯定是不会有重复的。也就是说dp[i][j]，可以拆分为：[i,j]区间里以字符k为最外层配对的回文数的和，其中k分别为'a','b','c','d'.

        dp[i][j]里以'a'为最外层配对的回文数怎么找呢？显然我们要把这个最外层配对找到，这是一定可以确定下来的。如果我们要找到i后面第一个为'a'的位置m，以及j前面第一个为'a'的位置n，并且m<n，显然就有递推的关系dp[i][j] += dp[m+1][n-1]，可以这么理解：拔掉了最外层的配对'a'，里面还有多少个回文数，就说明dp[i][j]有多少以'a'为最外层配对的回文数。同理我们可以找其他最外层配对的情况，也就是'b','c','d'。所以递推关系是：

        for (int k=0; k<4; k++)
        {
        if (next[i][k]<prev[j][k])  //小于号保证了这个配对一定是成双的
            dp[i][j] += dp[next[i][k]+1][prev[j][k]-1];
        }
        其中next[i][k]定义了i位置之后第一个为字符k的位置（可以是i本身），prev[j][k]定义了j位置之前第一个为字符k的位置（可以是j本身）。这两个next和prev数组可以预处理得到。

        上面的递推关系还有一个疏漏。如果[i,j]区间内没有以字符k的配对，但是有单独的一个字符k存在。这种情况也需要加进dp[i][j]的统计中去。所以完整的递推关系：

        for (int k=0; k<4; k++)
        {
        if (next[i][k]<prev[j][k])  //小于号保证了这个配对一定是成双的
            dp[i][j] += dp[next[i][k]+1][prev[j][k]-1];
        if (next[i][k]<=j)
            dp[i][j] += 1;
        }
        以上的语句再加上最外层对于i,j的大循环，即是完整的代码。最终的答案是dp[0][N-1]。

        本题其实也可以通过定义dp[i][j][k]三维状态数组来实现这个算法，最终输出是sum{k=0,1,2,3} dp[0][N-1][k]。但是空间上可能无法通过。
    */
    int countPalindromicSubsequences(string S) 
    {
        int N = S.size();
        auto dp = vector<vector<int>>(N,vector<int>(N,0));
        
        auto first = vector<vector<int>>(N,vector<int>(4,-1));
        auto last = vector<vector<int>>(N,vector<int>(4,-1));
        
        int MOD = 1e9+7;
        
        for (int len = 1; len<=N; len++)
            for (int i=0; i+len<=N; i++)
            {
                int j = i+len-1;
                
                if (first[i][S[j]-'a']==-1)
                    first[i][S[j]-'a'] = j;
                else
                    last[i][S[j]-'a'] = j;
                
                for (int k=0; k<4; k++)
                {
                    if (last[i][k]>=0)                    
                        dp[i][j]+=dp[first[i][k]+1][last[i][k]-1]+1;                    
                    if (first[i][k]>=0)
                        dp[i][j]+=1;                    
                    dp[i][j]%=MOD;
                }
            }
        
        return dp[0][N-1];
    }

    int countPalindromicSubsequences(string S) 
    {
        int M = 1e9+7;
        int N = S.size();
        auto next = vector<vector<int>>(N,vector<int>(4,N));
        auto prev = vector<vector<int>>(N,vector<int>(4,-1));        
        auto dp = vector<vector<int>>(N,vector<int>(N,0));  
        
        for (int k=0; k<4; k++)
        {
            int i = 0;
            for (int j=0; j<N; j++)
            {
                if (S[j]!='a'+k) continue;
                while (i<=j)
                {
                    next[i][k] = j;
                    i++;
                }
            }
        }
        
        for (int k=0; k<4; k++)
        {
            int i = N-1;
            for (int j=N-1; j>=0; j--)
            {
                if (S[j]!='a'+k) continue;
                while (i>=j)
                {
                    prev[i][k] = j;
                    i--;
                }
            }
        }
        
        for (int i=0; i<N; i++) dp[i][i]=1;
        
        for (int len = 2; len<=N; len++)
            for (int i=0; i<=N-len; i++)
            {
                int j = i+len-1;
                for (int k=0; k<4; k++)
                {
                    if (next[i][k]<prev[j][k])
                        dp[i][j]+=dp[next[i][k]+1][prev[j][k]-1]+1;
                    if (next[i][k]<=j)
                        dp[i][j]+=1;
                    dp[i][j] = dp[i][j]%M;
                }
            }
        return dp[0][N-1];
    }

public:
    
    int doit_dp_bottomup_best(string S) {
        int N = S.size();
        long MOD = 1e9 + 7;
        
        vector<vector<int>> T(N, vector<int>(N, 0));
        
        for (int i = 0; i < N; ++i)
            T[i][i] = 1;
        
        for (int gap = 1; gap < N; ++gap) {
            for (int i = 0, j = i+gap; j < N; ++i, ++j) {
                //Fill in T[i][j]
                if (S[i] != S[j]) {
                    T[i][j] = T[i+1][j] + T[i][j-1] - T[i+1][j-1];
                } else {
                    // Si, Si+1, ..., Sj-1, Sj
                    // Si+1, ...., Sj-1
                    T[i][j] = 2 * T[i+1][j-1];
                    
                    int l = i+1, r = j-1;
                    while(l <= r && S[l] != S[i])
                        l++;
                    while(l <= r && S[r] != S[i])
                        r--;
                    
                    if (l < r) {
                        // S(i, j) = b # # b . . b # # b where # != b, b + s[l+1][r-1] + b, duplicate with two outside b.
                        T[i][j] -= T[l+1][r-1];
                    }
                    
                    if (l == r) {
                        // S(i, j) = b # # b # # b where # != b, plus bb
                        T[i][j] += 1;
                    }
                    
                    if (l > r) {
                        // S(i, j) = b # # # b where # != b, plus b, bb
                        T[i][j] += 2;
                    }
                }
                T[i][j] = (T[i][j] + MOD)%MOD;
            }
        }
        
        return T[0][N-1];
    }

    
    int search(const string& S, int s, int e, vector<vector<int>>& dp) {
        
        int mod = 1e9+7;
        
        if (dp[s][e] != -1)
            return dp[s][e];
        
        if (e - s <= 2) {
            dp[s][e] = e - s;
            return e - s;
        }
        
        int cnt = 0;
        for (auto c : string("abcd")) {
            
            auto s1 = S.find(c, s);
            auto e1 = S.rfind(c, e-1);
            
            if (s1 == -1 && e1 == -1)
                continue;
            
            cnt += (e1 != s1) ? search(S, s1+1, e1, dp) + 2: 1;
        }
        
        dp[s][e] = cnt % mod;
        return cnt;
    }
    
    
    int doit_dp_dfs(string S) {
        
        auto N = S.length();
        vector<vector<int>> dp(N+1, vector<int>(N+1, -1));
        return search(S, 0, N, dp);
    }
};


void test_730_count_different_palindorme_subsequence() {
    
    CountPalindromicSubsequences().doit_dp_dfs(string("bccb"));
    
    CountPalindromicSubsequences().doit_dp_dfs(string("abcdabcdabcdabcdabcdabcdabcdabcddcbadcbadcbadcbadcbadcbadcbadcba"));
}
